provider/resources.go

```diff
@@ -1754,8 +1754,21 @@ func Provider() tfbridge.ProviderInfo {
 			},
 
 			// CloudRun V2 Resources
-			"google_cloud_run_v2_job":     {Tok: gcpResource(gcpCloudRunV2, "Job")},
-			"google_cloud_run_v2_service": {Tok: gcpResource(gcpCloudRunV2, "Service")},
+			"google_cloud_run_v2_job": {Tok: gcpResource(gcpCloudRunV2, "Job")},
+			"google_cloud_run_v2_service": {
+				Tok: gcpResource(gcpCloudRunV2, "Service"),
+				TransformFromState: func(_ context.Context, pMap resource.PropertyMap) (resource.PropertyMap, error) {
+					// `port` is an object type in the current version of this provider.
+					// But in a previous version, it was nested as a MaxItemsOne item, `ports`
+					// For correct updating, we extract the array item and re-write its content to the new `port` object.
+					if ports, ok := pMap["template"].ObjectValue()["containers"].ArrayValue()[0].ObjectValue()["ports"]; ok {
+						if ports.IsArray() {
+							pMap["template"].ObjectValue()["containers"].ArrayValue()[0].ObjectValue()["ports"] = ports.ArrayValue()[0]
+						}
+					}
```

Comment:
I would use path based changes here ([get](https://github.com/pulumi/pulumi/blob/770eb1e5450c1e88d0d4ec921b643b9bd565dc74/sdk/go/common/resource/properties_path.go#L136-L138), [set](https://github.com/pulumi/pulumi/blob/770eb1e5450c1e88d0d4ec921b643b9bd565dc74/sdk/go/common/resource/properties_path.go#L163-L165)), since they are less prone to panicing:

```suggestion
					ports := resource.PropertyPath{"template", "containers", 0, "ports"}
					if v, ok := ports.Get(pMap); ok && v.IsArray() {
						ports.Set(pMap, v.ArrayValue()[0])
					}
```

It's also a lot cleaner and less repetitive.

Reply:
I'm not actually familiar with these methods, so thank you for pointing them out.

However, your suggestion doesn't work as-is, because `.Get` takes a property value, not a property path. I'm genuinely confused how I'm supposed to use `.Get` without already having a property value.

Reply:
You can always encode a `resource.PropertyMap` into a `resource.PropertyValue`.

```go
pValue := resource.NewProperty(pMap)
if v, ok := ports.Get(pValue); ok && v.IsArray() {
	ports.Set(pValue v.ArrayValue()[0])
}

...

return pValue.ObjectValue()
```

Reply:
I'm really sorry; I've tried your suggestions - but ports.Get does not seem to work correctly:

```
pValue := resource.NewProperty(pMap)
containersPath := resource.PropertyPath{"template", "containers"}

if containers, ok := containersPath.Get(pValue); ok {
  contAry := containers.ArrayValue()
  for i, _ := range contAry {
	  portsPath := resource.PropertyPath{"template", "containers", fmt.Sprint(i), "ports"}
	  if val, ok := portsPath.Get(pValue); ok && val.IsArray() # this line is not finding the ports field
		  portsPath.Set(pValue, val.ArrayValue()[0])
	  }
  }
}
return pValue.ObjectValue(), nil
```

As a result, [ceeae57](https://github.com/pulumi/pulumi-gcp/pull/2355/commits/ceeae571f2d217108af468cef79b1f2d59794086) is failing in CI.

Given that due to containers being an array we'd have to break this up regardless, I would like to suggest we leave the explicit key lookups, despite the repetitiveness.

Reply:
`resource.PropertyPath` isn't a list of strings:

```go
// PropertyPath represents a path to a nested property. The path may be composed of strings (which access properties
// in ObjectProperty values) and integers (which access elements of ArrayProperty values).
type PropertyPath []interface{}
```
([src](https://github.com/pulumi/pulumi/blob/81847fedf30a1dfe7e7dabaabf15d6ca04a53bae/sdk/go/common/resource/properties_path.go#L28-L30))

`fmt.Sprint(i)` accesses the element of an object called `fmt.Sprint(i)`, it *is not* an index lookup. `i` is an index lookup.

I think that this would work:

```go
pValue := resource.NewProperty(pMap)
containersPath := resource.PropertyPath{"template", "containers"}

if containers, ok := containersPath.Get(pValue); ok {
	for i := range containers.ArrayValue() {
		portsPath := append(containersPath, i, "ports")
		if val, ok := portsPath.Get(pValue); ok && val.IsArray() {
			portsPath.Set(pValue, val.ArrayValue()[0])
		}
	}
}
return pValue.ObjectValue(), nil
```

I think that this would also work:

```go
pValue := resource.NewProperty(pMap)
containersPath := resource.PropertyPath{"template", "containers"}

if containers, ok := containersPath.Get(pValue); ok {
	arr := containers.ArrayValue()
	for i, v := range arr {
		portsPath := resource.PropertyPath{"ports"}
		if val, ok := portsPath.Get(v); ok && val.IsArray() {
			portsPath.Set(arr[i], val.ArrayValue()[0])
		}
	}
}
return pValue.ObjectValue(), nil
```